feat: 🎸 search service with tests

Galeaf11 opened the pull request:
✅ Closes: #45

Galeaf11 committed:
feat: 🎸 search service with tests

✅ Closes: #45

Galeaf11 committed:
feat: 🎸 fix from comments

fix tz to UTC and calculation in seconds

Galeaf11 committed:
feat: fix from comments
fix tz to UTC and calculation in seconds

kostysh commented on test/spaceSearch.spec.ts:
```diff
@@ -0,0 +1,330 @@
+  }
+
+  it('check search', async () => {
+    await init();
```
test case initialization task should be started from `before(async () => {});`
Tests can be grouped using `describe`. Each group of tests can have its own `before` initializer

Galeaf11 replied:
fixed

Galeaf11 committed:
feat: change init method to beforeEach

Galeaf11 committed:
feat: add facility timezone

convert all date-time to utc
change numRequired rule from days to seconds
add facility create to test

Galeaf11 committed:
feat: update space-models version

mfw78 commented on src/services/SearchService.ts:
```diff
@@ -1,7 +1,281 @@
+import ApiError from '../exceptions/ApiError';
+import { convertDaysToSeconds } from '../utils';
 
 export class SearchService {
```
In future, maybe it would be best to refactor the `ask` into the `search` so that the `SearchService` can be initialized per faciltiy with the connectors etc retained in memory to reduce initialization / query time.

Galeaf11 replied:
The database connector is a singleton, I think that this will not affect performance

mfw78 commented on src/services/SearchService.ts:
```diff
@@ -1,7 +1,281 @@
+    this.checkInTime =
+      this.facility.policies?.checkInTimeOneof.oneofKind === 'checkInTime'
+        ? this.facility.policies.checkInTimeOneof.checkInTime
+        : 'utc';
```
Does this not cause an issue if the `checkInTime` is undefined in the metadata? This will cause issues later when 'utc' is attempted to be cast to to a time. Is there any check in the tests for this?

Galeaf11 replied:
My bad, replace with '1500' and left todo '//todo think about default checkin time'

mfw78 replied:
Please raise this as an issue as an (sane defaults) so we can track it.

mfw78 commented on src/services/SearchService.ts:
```diff
@@ -1,7 +1,281 @@
+      throw ApiError.BadRequest('invalid dates in ask');
+    }
+
+    const checkIn = DateTime.fromFormat(this.checkInTime, 'hhmm', {
```
As this is processing _only dates_ (no times required), this should be handled just using dates, no casting to UTC or what not. Is this possible with this library?

Galeaf11 replied:
This is done for the convenience of further calculations, we calculate immediately in utc time,then we look at how much is left directly before the check-in

mfw78 commented on src/services/SearchService.ts:
```diff
@@ -1,7 +1,281 @@
+      spaceRuleRepository
+    )) as NoticeRequiredRule;
+
+    const firstDay = dates[0];
```
This variable should be `checkIn` as that's when it's proposed they want to `checkIn`

Galeaf11 replied:
fixed

mfw78 commented on src/services/SearchService.ts:
```diff
@@ -1,7 +1,281 @@
+    const firstDay = dates[0];
+
+    const checkIn = DateTime.now().setZone('utc');
+    const interval = Interval.fromDateTimes(checkIn, firstDay).toDuration([
```
In here, I'd recommend replacing `firstDay` with `DateTime.now().setZone('utc')` or in lieu of this, use a variable name like `now`.

Galeaf11 replied:
fixed

mfw78 commented on src/services/SearchService.ts:
```diff
@@ -1,7 +1,281 @@
+      //rule not exist
+    }
+
+    return (minLOS === 0 || lOS > minLOS) && (maxLOS === 0 || lOS < maxLOS);
```
`minLOS === 0` and `maxLOS === 0` are redundant due to initialisation.

Galeaf11 replied:
fixed

mfw78 commented on src/services/SearchService.ts:
```diff
@@ -1,7 +1,281 @@
+        ? space.maxNumberOfAdultOccupantsOneof.maxNumberOfAdultOccupants
+        : 0;
+
+    const numOfChildren =
```
Renaming to `maxNumOfChildren` to avoid confusing with `ask`

Galeaf11 replied:
fixed

mfw78 commented on src/services/SearchService.ts:
```diff
@@ -1,7 +1,281 @@
+  }
+
+  private checkSuitableQuantity(space: Space) {
+    const numOfAdults =
```
Renaming to `maxNumOfAdults` to avoid confusing with `ask`

Galeaf11 replied:
fixed

mfw78 commented on src/services/SearchService.ts:
```diff
@@ -1,7 +1,281 @@
+          formattedDate
+        );
+
+        if (available.numSpaces === 0) {
```
How about the case where intentionally the `numSpaces` is set to `0`, ie. an accommodation provider rams the `numSpaces available to 0 in order to "block" a date. If they do this, `numSpaces === 0` and then it will be assumed that the spaces available are actually the `default` number of spaces.

I think that this would require some editing of the repository, that if the daily override isn't found, instead of return a `numSpaces = 0` object, it should return undefined and this should be dealt with in the logic here (ie. if `!available.numSpaces` then `available = defaultAvailable` (as a `numSpaces = 0` is a legitimate value as well).

Please add a test case for this (ie. `numSpaces = 0` being set as a daily override to simulate "blocking" a space).
